WIP: TASK: Cleanup Node and TraversableNode interfaces

bwaidelich opened the pull request:
* Introduce `TraversableNodes` as replacement for `array<TraversableNodeInterface>`
* Replace nullable param and return types where possible
* Add missing (return) type annotations

skurfuerst commented:
@bwaidelich  I am currently very unsure if updating the *old* NodeInterface is a good idea. I would suggest to *remove* this from the change again, as it will be way more breaking than otherwise. What do you think?

I like the iterators etc :)

bwaidelich commented:
> I am currently very unsure if updating the old NodeInterface is a good idea

@skurfuerst you are absolutely right, my inner boy scout ran riot :)
I'll revert that and try to keep the type hints only for the new implementations.
I also want to get rid of some `try/catch` control flow hacks

bwaidelich commented:
I have reverted parts of the change and cleaned it up a little.
This is the corresponding change in the ES CR packages: https://github.com/neos/contentrepository-development-collection/pull/28
